File: app/src/main/java/d/d/meshenger/Event.java
package d.d.meshenger;

import org.json.JSONException;
import org.json.JSONObject;

import java.net.InetAddress;
import java.util.Date;

import d.d.meshenger.call.DirectRTCClient.CallDirection;


public class Event {
    public static final String TAG = "Event";

    public enum CallType {
        UNKNOWN,
        ACCEPTED,
        DECLINED,
        MISSED,
        ERROR
    };

    byte[] publicKey;
    String address; // may be null in case the call attempt failed
	CallDirection callDirection;
    CallType callType;
    Date date;

    public Event(byte[] publicKey, String address, CallDirection callDirection, CallType callType, Date date) {
        this.publicKey = publicKey;
        this.address = address;
        this.callDirection = callDirection;
        this.callType = callType;
        this.date = date;
    }

    public boolean isMissedCall() {
        return callDirection == CallDirection.INCOMING && (callType != CallType.ACCEPTED);
    }

    private static CallDirection callDirectionFromString(String value) {
        if (value.equals("INCOMING")) {
            return CallDirection.INCOMING;
        } else {
            return CallDirection.OUTGOING;
        }
    }

    private static String callDirectionToString(CallDirection value) {
        if (value == CallDirection.INCOMING) {
            return "INCOMING";
        } else {
            return "OUTGOING";
        }
    }

    private static String callTypeToString(CallType value) {
        switch (value) {
            case UNKNOWN: return "UNKNOWN";
            case ACCEPTED: return "ACCEPTED";
            case DECLINED: return "DECLINED";
            case MISSED: return "MISSED";
            case ERROR: return "ERROR";
            default:
                Log.w(TAG, "Invalid call type: " + value);
                return "UNKNOWN";
        }
    }

    private static CallType callTypeFromString(String value) {
        switch (value) {
            case "UNKNOWN": return CallType.UNKNOWN;
            case "ACCEPTED": return CallType.ACCEPTED;
            case "DECLINED": return CallType.DECLINED;
            case "MISSED": return CallType.MISSED;
            case "ERROR": return CallType.ERROR;
            default:
                Log.w(TAG, "Invalid call type: " + value);
                return CallType.UNKNOWN;
        }
    }

    public static JSONObject toJSON(Event event) throws JSONException {
        JSONObject obj = new JSONObject();
        obj.put("public_key", Utils.byteArrayToHexString(event.publicKey));
        obj.put("address", event.address);
        obj.put("call_direction", callDirectionToString(event.callDirection));
        obj.put("call_type", callTypeToString(event.callType));
        obj.put("date", String.valueOf(event.date.getTime()));
        return obj;
    }

    public static Event fromJSON(JSONObject obj) throws JSONException {
        byte[] publicKey = Utils.hexStringToByteArray(obj.getString("public_key"));
        String address = obj.getString("address");
        CallType callType = callTypeFromString(obj.getString("call_type"));
        CallDirection callDirection = callDirectionFromString(obj.getString("call_direction"));
        Date date = new Date(Long.parseLong(obj.getString("date"), 10));
        return new Event(publicKey, address, callDirection, callType, date);
    }
}

File: app/src/main/java/d/d/meshenger/Event.kt
package d.d.meshenger

import org.json.JSONException
import org.json.JSONObject
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.*

class Event(
    val publicKey: ByteArray, val address: InetSocketAddress?, val type: Type, val date: Date
) {
    enum class Type {
        UNKNOWN,
        OUTGOING_UNKNOWN, OUTGOING_ACCEPTED, OUTGOING_DECLINED, OUTGOING_MISSED, OUTGOING_ERROR,
        INCOMING_UNKNOWN, INCOMING_ACCEPTED, INCOMING_DECLINED, INCOMING_MISSED, INCOMING_ERROR
    }

    fun createUnknownContact(name: String): Contact {
        val addresses = mutableListOf<String>()
        val address = address?.address
        if (address != null) {
            // extract MAC address if possible
            val mac = AddressUtils.extractMacAddress(address)
            if (mac != null && mac.size == 6) {
                addresses.add(
                    "%02X:%02X:%02X:%02X:%02X:%02X".format(
                        mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]
                    )
                )
            } else {
                addresses.add(address.toString())
            }
        }
        return Contact(name, publicKey, addresses)
    }

    fun isMissedCall(): Boolean {
        return when (type) {
                Type.UNKNOWN -> false
                Type.OUTGOING_UNKNOWN -> false
                Type.OUTGOING_ACCEPTED -> false
                Type.OUTGOING_DECLINED -> false
                Type.OUTGOING_MISSED -> false
                Type.OUTGOING_ERROR -> false
                Type.INCOMING_UNKNOWN -> true
                Type.INCOMING_ACCEPTED -> false
                Type.INCOMING_DECLINED -> true
                Type.INCOMING_MISSED -> true
                Type.INCOMING_ERROR -> true
            }
    }

    companion object {
        private fun eventTypeToString(value: Type): String {
            return when (value) {
                Type.UNKNOWN -> "UNKNOWN"
                Type.OUTGOING_UNKNOWN -> "OUTGOING_UNKNOWN"
                Type.OUTGOING_ACCEPTED -> "OUTGOING_ACCEPTED"
                Type.OUTGOING_DECLINED -> "OUTGOING_DECLINED"
                Type.OUTGOING_MISSED -> "OUTGOING_MISSED"
                Type.OUTGOING_ERROR -> "OUTGOING_ERROR"
                Type.INCOMING_UNKNOWN -> "INCOMING_UNKNOWN"
                Type.INCOMING_ACCEPTED -> "INCOMING_ACCEPTED"
                Type.INCOMING_DECLINED -> "INCOMING_DECLINED"
                Type.INCOMING_MISSED -> "INCOMING_MISSED"
                Type.INCOMING_ERROR -> "INCOMING_ERROR"
            }
        }

        private fun eventTypeFromString(value: String): Type {
            return when (value) {
                "UNKNOWN" -> Type.UNKNOWN
                "OUTGOING_UNKNOWN" -> Type.OUTGOING_UNKNOWN
                "OUTGOING_ACCEPTED" -> Type.OUTGOING_ACCEPTED
                "OUTGOING_DECLINED" -> Type.OUTGOING_DECLINED
                "OUTGOING_MISSED" -> Type.OUTGOING_MISSED
                "OUTGOING_ERROR" -> Type.OUTGOING_ERROR
                "INCOMING_UNKNOWN" -> Type.INCOMING_UNKNOWN
                "INCOMING_ACCEPTED" -> Type.INCOMING_ACCEPTED
                "INCOMING_DECLINED" -> Type.INCOMING_DECLINED
                "INCOMING_MISSED" -> Type.INCOMING_MISSED
                "INCOMING_ERROR" -> Type.INCOMING_ERROR
                else -> {
                    Log.w(this, "Invalid call event type: $value")
                    Type.OUTGOING_UNKNOWN
                }
            }
        }

        @Throws(JSONException::class)
        fun toJSON(event: Event): JSONObject {
            val obj = JSONObject()
            obj.put("public_key", Utils.byteArrayToHexString(event.publicKey))
            obj.put("address", AddressUtils.inetSocketAddressToString(event.address))
            obj.put("type", eventTypeToString(event.type))
            obj.put("date", event.date.time.toString())
            return obj
        }

        @Throws(JSONException::class)
        fun fromJSON(obj: JSONObject): Event {
            val publicKey = Utils.hexStringToByteArray(obj.getString("public_key"))!!
            val address = AddressUtils.stringToInetSocketAddress(obj.optString("address"), MainService.serverPort.toUShort())
            val type = eventTypeFromString(obj.getString("type"))
            val date = Date(obj.getString("date").toLong(10))
            return Event(publicKey, address, type, date)
        }
    }
}